src/Infrastructure/LeanCode.AI.McpServer/McpServerBuilderExtensions.cs

```diff
@@ -0,0 +1,133 @@
+using LeanCode.Contracts;
+using LeanCode.CQRS.AspNetCore.Registration;
+using LeanCode.CQRS.Execution;
+using Microsoft.Extensions.DependencyInjection;
+using ModelContextProtocol.Server;
+
+namespace LeanCode.AI.McpServer;
+
+public static class McpServerBuilderExtensions
+{
+    private static Type GetResultType(Type cqrsObjectType, CQRSObjectKind objectKind)
+    {
+        return objectKind switch
+        {
+            CQRSObjectKind.Command => typeof(CommandResult),
+            CQRSObjectKind.Query => cqrsObjectType
+                .GetInterfaces()
+                .Single(i => i.IsConstructedGenericType && i.GetGenericTypeDefinition() == typeof(IQuery<>))
+                .GetGenericArguments()
+                .Single(),
+            CQRSObjectKind.Operation => cqrsObjectType
+                .GetInterfaces()
+                .Single(i => i.IsConstructedGenericType && i.GetGenericTypeDefinition() == typeof(IOperation<>))
+                .GetGenericArguments()
+                .Single(),
+            _ => throw new ArgumentOutOfRangeException(nameof(objectKind)),
+        };
+    }
+
+    private static IMcpServerBuilder WithToolsFromContracts(
+        this IMcpServerBuilder builder,
+        CQRSObjectKind objectKind,
+        Func<Type, Type, Type, Type> typeConstructor,
+        TypesCatalog contracts,
+        Type executorType,
+        CQRSAIFunction.JsonSchemaSource jsonSchemaSource = default,
+        JsonSerializerOptions? serializerOptions = null
+    )
+    {
+        var tools = new List<McpServerTool>();
+
+        foreach (var (contract, kind) in CQRSObjectCollection.EnumerateCQRSObjects(contracts))
+        {
+            if (kind != objectKind || contract.GetCustomAttribute<McpToolAttribute>() is null)
+            {
+                continue;
+            }
+
+            var type = typeConstructor(executorType, contract, GetResultType(contract, objectKind));
+            var function = (CQRSAIFunction)Activator.CreateInstance(type, [jsonSchemaSource])!;
+
+            tools.Add(
+                McpServerTool.Create(
+                    function,
+                    new()
+                    {
+                        Name = function.Name,
+                        Description = function.Description,
+                        Title = function.Title,
+                        Destructive = function.ToolHints.HasFlag(McpToolHints.Destructive),
+                        Idempotent = function.ToolHints.HasFlag(McpToolHints.Idempotent),
+                        OpenWorld = function.ToolHints.HasFlag(McpToolHints.OpenWorld),
+                        ReadOnly = function.ToolHints.HasFlag(McpToolHints.ReadOnly),
+                        // TODO: UseStructuredContent = true,
+                        SerializerOptions = serializerOptions,
+                    }
+                )
+            );
+        }
+
+        return builder.WithTools(tools);
+    }
+
+    public static IMcpServerBuilder WithToolsFromCommands(
```

Comment:
What's the `executorType`?

Reply:
It configures what subtype of `HttpCommandsExecutor` should be retrieved from DI to execute the remote CQRS call.

But your comment made me realized I can move it to generic type parameters, so I'll do just that.